modules/core/src/test/java/org/apache/ignite/internal/IgniteClientCheckClusterGroupLocalIdAfterReconnect.java

```diff
@@ -0,0 +1,139 @@
+        /** */
+        private External() {}
+
+        /** {@inheritDoc} */
+        @Override public void writeExternal(ObjectOutput out) throws IOException {}
+
+        /** {@inheritDoc} */
+        @Override public void readExternal(ObjectInput in) throws IOException, ClassNotFoundException {}
+    }
+
+    /** {@inheritDoc} */
+    @Override protected IgniteConfiguration getConfiguration(String igniteInstanceName) throws Exception {
+        IgniteConfiguration cfg = super.getConfiguration(igniteInstanceName);
+
+        cfg.setFailureDetectionTimeout(5000);
+        cfg.setClientFailureDetectionTimeout(5000);
+
+        return cfg;
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void afterTest() throws Exception {
+        super.afterTest();
+
+        stopAllGrids();
+    }
+
+    /**
+     * Test checks that local id in cluster group was change and client
+     * will be able to send the message to itself after reconnect.
+     */
+    @Test
+    public void testClusterGroupLocalIdAfterClientReconnect() throws Exception {
+        Ignite server = startGrid(0);
+
+        Ignite client = startClientGrid(1);
+
+        UUID clientId = client.cluster().node().id();
+
+        ClusterGroup cg1 = client.cluster().forLocal();
+
+        assertNotNull("Local client ID is different with local ClusterGroup node id. ", cg1.node(clientId));
+
+        // check sending messages is possible while connected
+        IgniteMessaging messaging = client.message(client.cluster().forLocal());
+
+        CountDownLatch topicSignal = new CountDownLatch(1);
+
+        messaging.localListen("topic", (IgniteBiPredicate<UUID, Object>) (uuid, n) -> {
+            topicSignal.countDown();
+
+            return true;
+        });
+
+        messaging.send("topic", new External());
+
+        CountDownLatch discSignal = new CountDownLatch(1);
+
+        client.events().localListen((IgnitePredicate<DiscoveryEvent>) evt -> {
+            discSignal.countDown();
+
+            return true;
+        }, EventType.EVT_CLIENT_NODE_DISCONNECTED);
+
+        server.close();
+
+        assertTrue("client did not disconnect", discSignal.await(LATCH_TIMEOUT, TimeUnit.SECONDS));
+
+        startGrid(0);
+
+        // wait for client reconnect
+        IgniteFuture future = client.cluster().clientReconnectFuture();
+
+        assertNotNull(future);
+
+        future.get(20_000);   // throws if times out
+
+        ClusterGroup cg2 = client.cluster().forLocal();
+
+        UUID newClientId = client.cluster().localNode().id();
+
+        assertNotNull("Local client ID wasn't changed for local ClusterGroup.", cg2.node(newClientId));
+
+        awaitPartitionMapExchange();
+
+        // check sending messages is possible after reconnecting
+        messaging = client.message(client.cluster().forLocal());
+
+        messaging.send("topic", new External());
+
+        assertTrue("Message wasn't received", topicSignal.await(LATCH_TIMEOUT, TimeUnit.SECONDS));
```

Comment:
Latch here is already zero, right? Even without the last send?

Reply:
CountDownLatch now has a count - 2